flaml/automl/time_series/tft.py

```diff
@@ -0,0 +1,193 @@
+
+        from pytorch_forecasting import TimeSeriesDataSet
+        from pytorch_forecasting.data import GroupNormalizer
+
+        self.group_ids = kwargs["group_ids"].copy()
+        training = TimeSeriesDataSet(
+            self.data[lambda x: x.time_idx <= training_cutoff],
+            time_idx="time_idx",
+            target=X_train.target_names[0],
+            group_ids=self.group_ids,
+            min_encoder_length=kwargs.get(
+                "min_encoder_length", self.max_encoder_length // 2
+            ),  # keep encoder length long (as it is in the validation set)
+            max_encoder_length=self.max_encoder_length,
+            min_prediction_length=1,
+            max_prediction_length=max_prediction_length,
+            static_categoricals=kwargs.get("static_categoricals", []),
+            static_reals=kwargs.get("static_reals", []),
+            time_varying_known_categoricals=kwargs.get(
+                "time_varying_known_categoricals", []
+            ),
+            time_varying_known_reals=kwargs.get("time_varying_known_reals", []),
+            time_varying_unknown_categoricals=kwargs.get(
+                "time_varying_unknown_categoricals", []
+            ),
+            time_varying_unknown_reals=kwargs.get("time_varying_unknown_reals", []),
+            variable_groups=kwargs.get(
+                "variable_groups", {}
+            ),  # group of categorical variables can be treated as one variable
+            lags=kwargs.get("lags", {}),
+            target_normalizer=GroupNormalizer(
+                groups=kwargs["group_ids"], transformation="softplus"
+            ),  # use softplus and normalize by group
+            add_relative_time_idx=True,
+            add_target_scales=True,
+            add_encoder_length=True,
+        )
+
+        # create validation set (predict=True) which means to predict the last max_prediction_length points in time
+        # for each series
+        validation = TimeSeriesDataSet.from_dataset(
+            training, self.data, predict=True, stop_randomization=True
+        )
+
+        # create dataloaders for model
+        batch_size = kwargs.get("batch_size", 64)
+        train_dataloader = training.to_dataloader(
+            train=True, batch_size=batch_size, num_workers=0
+        )
+        val_dataloader = validation.to_dataloader(
+            train=False, batch_size=batch_size * 10, num_workers=0
+        )
+
+        return training, train_dataloader, val_dataloader
+
+    def fit(self, X_train, y_train, budget=None, **kwargs):
+        import warnings
+        import pytorch_lightning as pl
+        import torch
+        from pytorch_forecasting import TemporalFusionTransformer
+        from pytorch_forecasting.metrics import QuantileLoss
+        from pytorch_lightning.callbacks import EarlyStopping, LearningRateMonitor
+        from pytorch_lightning.loggers import TensorBoardLogger
+
+        # a bit of monkey patching to fix the MacOS test
```

Comment:
As all this function appeared to do is some plotting during a fit (which we don't do for other models) and somehow caused a fail on MacOS, with no way I could find to turn it off,  I monkey-patch it here to do nothing instead :)

Reply:
There is actually a way to turn it off. See https://github.com/microsoft/FLAML/blob/cdc4201188565e74507c026f86abe540ade4cb72/flaml/automl/model.py#L2275

``` python
        def _fit(log):
            default_trainer_kwargs = dict(
                gpus=kwargs.get("gpu_per_trial", [0])
                if torch.cuda.is_available()
                else None,
                max_epochs=max_epochs,
                gradient_clip_val=gradient_clip_val,
                callbacks=[LearningRateMonitor(), early_stop_callback]
                if log
                else [early_stop_callback],
                logger=log,
            )
            trainer = pl.Trainer(
                **default_trainer_kwargs,
            )
            tft = TemporalFusionTransformer.from_dataset(
                training,
                **params,
                lstm_layers=2,  # 2 is mostly optimal according to documentation
                output_size=7,  # 7 quantiles by default
                loss=QuantileLoss(),
                log_interval=10 if log else 0,
                # uncomment for learning rate finder and otherwise, e.g. to 10 for logging every 10 batches
                reduce_on_plateau_patience=4,
            )
            # fit network
            trainer.fit(
                tft,
                train_dataloaders=train_dataloader,
                val_dataloaders=val_dataloader,
            )
            return trainer
        trainer = _fit(log=False)
```